dll/win32/shell32/shlexec.cpp

```diff
@@ -458,6 +458,132 @@ static HRESULT SHELL_GetPathFromIDListForExecuteW(LPCITEMIDLIST pidl, LPWSTR psz
+                }
+                SendDlgItemMessage(hwnd, IDC_RUNAS_NAME, CB_SETCURSEL, 0, 0);
+                return TRUE;
+            }
+            break;
+        case WM_COMMAND:
+            switch(LOWORD(wParam))
+            {
+                case IDCANCEL:
+                    EndDialog(hwnd, 1);
+                    break;
+                case IDOK:
+                    {
+                        pData->Safer = SendDlgItemMessage(hwnd, IDC_RUNAS_SAFER,
+                                                          BM_GETCHECK, 0, 0);
+                        SendDlgItemMessage(hwnd, IDC_RUNAS_NAME, WM_GETTEXT,
+                                           _countof(RUNASDLGDATA::NameBuffer), (LPARAM)pData->NameBuffer);
+                        SendDlgItemMessage(hwnd, IDC_RUNAS_PASS, WM_GETTEXT,
+                                           _countof(RUNASDLGDATA::Pass), (LPARAM)pData->Pass);
+                        UINT other = SendDlgItemMessage(hwnd, IDC_RUNAS_OTHER,
+                                                        BM_GETCHECK, 0, 0) != 0;
+                        pData->Name = pData->NameBuffer;
+                        pData->Domain = strstrW(pData->Name, L"\\");
+                        if (pData->Domain)
+                        {
+                            LPWSTR tmp = pData->Domain + 1;
+                            pData->Domain[0] = UNICODE_NULL;
+                            pData->Domain = pData->Name;
+                            pData->Name = tmp;
+                        }
+                        if (other)
+                        {
+                            HANDLE token;
+                            if (!LogonUser(pData->Name, pData->Domain, pData->Pass,
+                                           LOGON32_LOGON_INTERACTIVE, LOGON32_PROVIDER_DEFAULT,
+                                           &token))
+                                return SHELL_ErrorBox(hwnd);
+                            CloseHandle(token);
+                            other |= (GetKeyState(VK_SHIFT) & 0x8000);
+                        }
+                        EndDialog(hwnd, 2 + other);
+                    }
+                    break;
+                case IDC_RUNAS_BROWSE:
+                    return SHELL_ErrorBox(hwnd, ERROR_CALL_NOT_IMPLEMENTED); // TODO
+            }
+            break;
+    }
+    return FALSE;
+}
+
+static HRESULT SHELL_PromptAndRunProcessAs(HWND hwnd, LPWSTR Cmd, DWORD CreationFlags,
+                                           LPWSTR Env, LPCWSTR Dir,
+                                           STARTUPINFOW *pSI, PROCESS_INFORMATION *pPI)
+{
+    RUNASDLGDATA data;
+    INT_PTR dlgret = DialogBoxParamW(shell32_hInstance, MAKEINTRESOURCEW(IDD_RUN_AS),
+                                     SHELL_GetUsableDialogOwner(hwnd), RunAsDlgProc, (LPARAM)&data);
+    if (dlgret > 2)
+    {
+        UINT logonFlags = (dlgret & 0x8000) ? LOGON_NETCREDENTIALS_ONLY : LOGON_WITH_PROFILE;
```

Comment:
Is this "0x8000" the keypress flag being set by the GetKeyState() call from the dialog-proc?

Reply:
Yes